### PR TITLE
[Improvement-3369][api] Introduce taskrecord, udffunc, workflowlineage and workergroup service interface for clear code

### LINKED ISSUE #3369: [Improvement][api] Introduce interface about api module service
**Describe the question**
Introduce interface about api module service for clearly service module.

**What are the current deficiencies and the benefits of improvement**
- A clear and concise service will help contributor or user to ready the code and locate the function.

**Which version of DolphinScheduler:**
 -[dev]

**Describe alternatives you've considered**
Introduce service interface

### PR DESCRIPTION
**Pull request by chengshiwen**
## What is the purpose of the pull request

*Introduce taskrecord, udffunc, workflowlineage and workergroup service interface for clear code and close pr #3445 which has conflicts and not updated for a long time*

Compared to pr #3445 with **2 vulnerabilities and 18 code smells** half year ago, there are only **26 code smells** in this pr.

This closes #3369

## Brief change log

  - *Add TaskRecordServiceeImpl and change TaskRecordService to interface*
  - *Add UdfFuncServiceImpl and change UdfFuncService to interface*
  - *Add WorkFlowLineageServiceImpl and change WorkFlowLineageService to interface*
  - *Add WorkerGroupServiceImpl and change WorkerGroupService to interface*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

### CODE REVIEW

**Comment by chengshiwen** [comment]
@CalvinKirs Now only 5 code smells
image

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+     *
+     * @param loginUser login user
+     * @param pageNo page number
+     * @param searchVal search value
+     * @param pageSize page size
+     * @return worker group list page
+     */
+    public Map<String, Object> queryAllGroupPaging(User loginUser, Integer pageNo, Integer pageSize, String searchVal) {
+        // list from index
+        Integer fromIndex = (pageNo - 1) * pageSize;
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java]: Modified to basic type

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                    return workerGroups;
+                } else {
+                    //ignore noNodeException return Default
+                    WorkerGroup wg = new WorkerGroup();
+                    wg.setName(DEFAULT_WORKER_GROUP);
+                    workerGroups.add(wg);
+                    return workerGroups;
+                }
+            } else {
+                throw e;
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java:152]: This approach doesn't look very good

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ * work flow lineage service impl
+ */
+@Service
+public class WorkFlowLineageServiceImpl extends BaseService implements WorkFlowLineageService {
+
+    @Autowired
+    private WorkFlowLineageMapper workFlowLineageMapper;
+
+    public Map<String, Object> queryWorkFlowLineageByName(String workFlowName, int projectId) {
+        Map<String, Object> result = new HashMap<>(5);
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java]: It doesn't make much sense to declare the initial size as 5

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        List<WorkFlowLineage> workFlowLineageList = workFlowLineageMapper.queryByName(workFlowName, projectId);
+        result.put(Constants.DATA_LIST, workFlowLineageList);
+        putMsg(result, Status.SUCCESS);
+        return result;
+    }
+
+    private List<WorkFlowRelation> getWorkFlowRelationRecursion(Set<Integer> ids, List<WorkFlowRelation> workFlowRelations,Set<Integer> sourceIds) {
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java]: This method may be better declared as void

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            for (int id : ids) {
+                idSet.remove(id);
+            }
+            if (!idSet.isEmpty()) {
+                workFlowLineageList.addAll(workFlowLineageMapper.queryByIds(idSet, projectId));
+            }
+        }
+
+        workFlowLists.put("workFlowList",workFlowLineageList);
```
Comment (CalvinKirs) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java]: It's best not to show magic numbers

**Review by CalvinKirs — CHANGES_REQUESTED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java (hunk 3 above)
Comment (chengshiwen) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java]: Has modified to 0

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java (hunk 1 above)
Comment (chengshiwen) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java]: Done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java (hunk 4 above)
Comment (chengshiwen) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java]: Done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java (hunk 5 above)
Comment (chengshiwen) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkFlowLineageServiceImpl.java]: Replaced with `Constants.WORKFLOW_LIST` and `Constants.WORKFLOW_RELATION_LIST`

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java (hunk 2 above)
Comment (chengshiwen) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkerGroupServiceImpl.java:152]: image
Exception 'org.apache.zookeeper.KeeperException.NoNodeException' is never thrown in the corresponding try block
Without looking at the code details, I can only simply modify this piece of code

**Review by CalvinKirs — APPROVED** [review]
LGTM

**Pull request merged**